File: test/org/cerion/symcalc/expression/function/list/FlattenTest.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.junit.Test;

import static org.junit.Assert.*;

public class FlattenTest {

    private static final ListExpr L1 = new ListExpr(IntegerNum.ONE, IntegerNum.TWO);
    private static final ListExpr L2 = new ListExpr(new IntegerNum(3), new IntegerNum(4));
    private static final ListExpr L3 = new ListExpr(new IntegerNum(5), new IntegerNum(6));

    @Test
    public void validateParameters() {
        Expr e = new Flatten(IntegerNum.ZERO);
        e = e.eval();
        assertTrue(e.isError());
    }

    @Test
    public void singleList() {
        Expr e = new Flatten(new ListExpr(IntegerNum.ZERO, L1)).eval();
        verify(new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO), e);

        e = new Flatten(new ListExpr(L1, IntegerNum.ZERO)).eval();
        verify(new ListExpr(IntegerNum.ONE, IntegerNum.TWO, IntegerNum.ZERO), e);
    }

    @Test
    public void nestedList() {
        ListExpr list = new ListExpr(IntegerNum.ZERO, L1, new ListExpr(L2,L3));
        ListExpr expected = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO,
                new IntegerNum(3), new IntegerNum(4), new IntegerNum(5), new IntegerNum(6));

        verify(expected, new Flatten(list).eval());
    }

    private void verify(ListExpr expected, Expr actual) {
        assertTrue(actual.isList());
        ListExpr l = (ListExpr)actual;

        assertEquals(expected.size(), l.size());
        assertEquals(expected, actual);
    }
}

File: test/expression/function/list/FlattenTest.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.number.IntegerNum
import org.junit.Test

import org.junit.Assert.*

class FlattenTest {

    @Test
    fun validateParameters() {
        var e: Expr = Flatten(IntegerNum.ZERO)
        e = e.eval()
        assertTrue(e.isError)
    }

    @Test
    fun singleList() {
        var e = Flatten(ListExpr(IntegerNum.ZERO, L1)).eval()
        verify(ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO), e)

        e = Flatten(ListExpr(L1, IntegerNum.ZERO)).eval()
        verify(ListExpr(IntegerNum.ONE, IntegerNum.TWO, IntegerNum.ZERO), e)
    }

    @Test
    fun nestedList() {
        val list = ListExpr(IntegerNum.ZERO, L1, ListExpr(L2, L3))
        val expected = ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO,
                IntegerNum(3), IntegerNum(4), IntegerNum(5), IntegerNum(6))

        verify(expected, Flatten(list).eval())
    }

    private fun verify(expected: ListExpr, actual: Expr) {
        assertTrue(actual.isList)
        val l = actual as ListExpr

        assertEquals(expected.size.toLong(), l.size.toLong())
        assertEquals(expected, actual)
    }

    companion object {
        private val L1 = ListExpr(IntegerNum.ONE, IntegerNum.TWO)
        private val L2 = ListExpr(IntegerNum(3), IntegerNum(4))
        private val L3 = ListExpr(IntegerNum(5), IntegerNum(6))
    }
}